### ISSUE TITLE
Constraint solver cannot find solution when union of "naked" type variables are used in parameter annotation

### ISSUE THREAD

**Issue by robertmaxton42** [microsoft/pyright#5768]
Suppose I define the functions

```python
from typing import TypeVar, Callable
S = TypeVar("S")
T = TypeVar("T")
U = TypeVar("U")

def accepts_bool(b: bool):
    ...

def accepts_int(i: int):
    ...

def mwe1(x: S | T, l1: Callable[..., S], l2: Callable[[S], U], l3: Callable[[T], U]) -> U:
    ...

def mwe2(x: int | bool):
    return mwe1(x, lambda y: 0, accepts_int, accepts_bool)
```
On the one hand, while unifying `S | T` to `int | bool` directly is ambiguous, since the lambda in the second argument returns an int, the second argument should disambiguate the assignments of `S` and `T` in favor of `S = int, T = bool`. On the other hand, I will discover that Pyright believes that the last argument to `mwe1` should also have type `(int->None)`, and `accepts_bool` is invalid. Also, if I replace the functions with lambdas `lambda i: i, lambda b: b`, Pyright will inform me that `i` and `b` have the inferred types `(bool | int)` and `int` respectively, where they should be simply `int` and `bool`. (Though I'm not sure how much this helps you given that lambdas are hard to annotate.)

My use case here is in writing type-guarded higher order functions; for example, I wanted to write the following wrapper to JAX's `jax.lax.cond`:

```python
Maybe : TypeAlias = T | Any             # A Maybe[T] might be a T. (Equivalent to Any but with clearer semantics.)
TypeGuardFunc : TypeAlias = Callable[Concatenate[Maybe[T], ...], TypeGuard[T]]

def typeguard_cond(
        x: S | T, typeguard: TypeGuardFunc[S], 
        true_fun: Callable[Concatenate[S, P], U], false_fun: Callable[Concatenate[T, P], V],
        *operands: P.args, **kwoperands: P.kwargs) -> U | V:
    return lax.cond(
        typeguard(x), 
        lambda z: true_fun(typing.cast(S, z), *operands, **kwoperands), 
        lambda z: false_fun(typing.cast(T, z), *operands, **kwoperands), x)
```

When I actually call that function, though, I end up with an error on the other (third argument) lambda: since the type of the bound variable is incorrectly inferred as `S | T` and not simply `S`, I can't use members/features that only apply to `S`. Thus the title.

**VS Code extension or command-line**
VS Code Pylance, pre-release version 2023.8.31.

**Comment by erictraut** [comment]
This isn't a type narrowing issue, as the title would suggest. It's a failure of the constraint solver to find a solution for `S` and `T` that meets all of the constraints. Any time you have a "naked" type variable in a union with another type, it creates ambiguities for the constraint solver. In this case, you have two such "naked" type variables in the same union, which is especially challenging.

I'll spend some time investigating whether it's feasible to fix this without adding undue complexity or performance overhead to the constraint solver, but I consider this an extreme edge case, so the answer might be "close as won't fix" in the end.

**Comment by erictraut** [comment]
Digging into this a bit more...

My first observation is that the type described by the annotation `int | bool` is the same as the type described by the annotation `int` because `bool` is a proper subtype of `int`. We can therefore simplify the code by replacing `int | bool` with just `int`. Second, the parameter `l1` adds no meaningful constraints, so it can be eliminated for purposes of demonstrating the issue. Third, the return types of the callables for `l2` and `l3` are also irrelevant, so `U` can be eliminated.

Here's a simplified repro:

```python
from typing import Any, TypeVar, Callable

S = TypeVar("S")
T = TypeVar("T")

def accepts_bool(b: bool) -> None: ...
def accepts_int(i: int) -> None: ...

def mwe1(x: S | T, l2: Callable[[S], Any], l3: Callable[[T], Any]) -> tuple[S, T]:
    ...

x = mwe1(0, accepts_int, accepts_bool)
reveal_type(x)  # Should be tuple[int, bool]
```

**Comment by erictraut** [comment]
I've added some additional heuristics that enable type checking of this code without error. The change will be in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.324, which I just published. It will also be included in a future release of pylance.
